### PR TITLE
fix: compose notification center delegates (MBL-2281)

### PR DESCRIPTION
**Pull request by Shahroz16**
## What changed

- retain one Customer.io `UNUserNotificationCenterDelegate` proxy and compose host or third-party delegates as identity-distinct peers
- preserve delegates assigned both before and after SDK initialization, including public Customer.io delegate instances
- fan out notification callbacks over a stable per-delivery snapshot and arbitrate asynchronous completion handlers exactly once
- preserve existing single-peer foreground presentation behavior, while unioning options across multiple implementing peers
- handle `nil` clears, duplicate assignments, reentrant assignments, concurrent installation, and bounded weak-peer cleanup

## Why

The current single-wrapper implementation replaces the previously retained delegate whenever another delegate is assigned. In an A-before-Customer.io, B-after-Customer.io sequence, A stops receiving `willPresent` and `didReceive`. This breaks coexistence with apps and notification SDKs that install delegates at different lifecycle points.

## Validation

- Xcode 26.6 `MessagingPushAPN` build and `Customer.io-Package` build-for-testing succeeded on iPhone 17 Pro, iOS 26.5 simulator
- 78 focused tests passed across the five affected suites
- four high-risk subclass-order, abandonment, and observer-reentry regressions passed in ten repeated runs (40/40)
- strict SwiftLint passed with 0 violations in 733 files
- `git diff --check`, pre-commit, and pre-push hooks passed
- fresh exact-byte adversarial review accepted with no P0-P3 findings after every reported lock, lifetime, observer, and transitive-delivery issue was corrected

Claude Code Opus was unavailable because its account spend limit was reached; no Claude verdict is claimed. The independent adversarial review was performed by a separate spawned reviewer and checked the final bytes before commit.

## Evidence boundary

These tests cover SDK composition, concurrency, ownership, and completion arbitration with simulator and host doubles. Real Apple weak-delegate behavior, arbitrary third-party swizzle ordering, and physical-device APN/FCM delivery remain for the cross-SDK runtime matrix in MBL-2233.

Linear: MBL-2281

---

> [!NOTE]
> **High Risk**
> Changes core push notification delegate installation, Objective-C swizzling, and callback fan-out/completion semantics that affect every push interaction and coexistence with third-party notification SDKs.
> 
> **Overview**
> **Customer.io now keeps one `UNUserNotificationCenter` proxy installed** and registers every other delegate assignment as a **peer** instead of replacing the previous wrapper. Delegates set before or after `MessagingPush.initialize()` (including other Customer.io delegate instances) keep receiving `willPresent`, `didReceive`, and `openSettings` through a stable per-delivery snapshot.
> 
> The swizzled `delegate` setter **registers peers** and reasserts the shared proxy when competing SDK swizzlers allow it; `nil` clears peers without uninstalling the proxy. Install logic moves to dedicated helpers with **`installLock`**, a **clear-generation** counter for racing nil captures, and bounded proxy reassertion when another setter wins.
> 
> **`CioNotificationCenterDelegate`** fans out to implementing peers, unions `willPresent` options across multiple peers (with the existing CIO fallback when none implement it), runs Customer.io open handling once, and uses new helpers (`PeerDeliveryCompletionAggregate`, delivery/peer guards, forwarding markers) so system completion handlers fire **exactly once** and reentrant or cyclic forwarding does not double-handle events.
> 
> Initialization calls **`installNotificationCenterDelegate(centerProvider:)`** without pre-wrapping the current delegate; the first-install path captures whatever is already on the center. Large new unit/integration test suites cover multi-peer behavior, concurrency, and swizzle edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6578209dc6b4cf0365e562272adf855e7f628eea. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by mahmoud-elmorabea — COMMENTED** [review]
- `DeliveryIdentifier` is `ObjectIdentifier(response)`, a reusable address: a peer that never calls its completion handler leaks `activeDeliveries`, and a later response at the same address is then treated as re-entrant and skips `handlePushResponse`. Lifecycle-independent, so it applies on Xcode 26 too.
- Unstack #1215 from this branch so both can land independently.

File: Sources/MessagingPush/Integration/CioNotificationCenterDelegate.swift [hunk 1]
```diff
@@ -57,34 +146,170 @@ open class CioNotificationCenterDelegate: NSObject, UNUserNotificationCenterDele
...
         withCompletionHandler completionHandler: @escaping () -> Void
     ) {
-        // Cast to concrete type since method was removed from protocol
-        if let implementation = messagingPush as? MessagingPush {
-            _ = implementation.userNotificationCenter(center, didReceive: response)
+        let deliveryIdentifier = DeliveryIdentifier.didReceive(ObjectIdentifier(response))
+        guard beginDelivery(deliveryIdentifier) else {
+            // The outer invocation already performed Customer.io handling and owns the peer aggregate.
+            completionHandler()
+            return
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply Sources/MessagingPush/Integration/CioNotificationCenterDelegate.swift:161]: Good catch on the abandoned-delivery identity lifetime. The aggregate now retains the exact notification/response while its guard is active and clears the guard on abandonment, preventing ObjectIdentifier reuse from poisoning a later delivery. A peer that retains its completion remains intentionally in flight; we preserve the existing no-timeout/no-synthesized-completion policy. Focused tests cover abandonment, successful completion, asynchronous retention, and repeated delivery, and Opus found no production blocker after one test-hardening follow-up.

**Comment by Shahroz16** [comment]
Both points are addressed:

- The current delivery guard retains the exact notification/response while its `ObjectIdentifier` is active. If all peers discard completion, aggregate deinit clears the guard before releasing that object, so its address cannot be reused while stale. If a peer retains completion forever, the existing no-timeout behavior intentionally retains both the guard and exact object, which also prevents address reuse. Focused Opus review found no production blocker; the success/deinit coverage it requested was added.
- #1215 is now unstacked. Its sole feature commit was rebased onto current `main`, force-updated with lease, and PR #1215 now targets `main`. #1211 never contained #1215 source changes.

Post-unstack validation for #1215: package test build, 24 coordinator tests, APN sample, regenerated CocoaPods FCM sample, format/lint, and diff check all pass.

**Comment by Shahroz16** [comment]
Both concerns are addressed. The delivery guard now retains the exact delivery object for as long as an async completion can remain outstanding, so another response cannot reuse that identity; abandonment clears the guard before the object is released. The intentional no-timeout behavior is documented and stress-tested. PR #1215 is now independently based on main and contains no #1211 implementation. Claude Opus found no production blocker in the corrected exact diff. Please review the current head.

**Review by mahmoud-elmorabea — APPROVED** [review]
